### PR TITLE
Port synapse.replication.tcp to async/await

### PR DESCRIPTION
**Pull request by erikjohnston**

### CODE REVIEW

**Comment by erikjohnston** [comment]
Yeah, agreed. I could have just documented the functions as returning `Awaitable`, but instead I've just gone and marked them all as `async`, as that felt saner.

File: synapse/replication/tcp/protocol.py [hunk 1]
```diff
@@ -376,7 +372,7 @@ def connectionLost(self, reason):
 
         self.on_connection_closed()
 
-    def on_connection_closed(self):
+    async def on_connection_closed(self):
```
Comment (richvdh) [inline synapse/replication/tcp/protocol.py]: this one isn't called the same way

File: synapse/replication/tcp/protocol.py [hunk 2]
```diff
@@ -572,27 +567,24 @@ class AbstractReplicationClientHandler(metaclass=abc.ABCMeta):
...
 
     @abc.abstractmethod
-    def on_position(self, stream_name, token):
+    async def on_position(self, stream_name, token):
         """Called when we get new position data."""
         raise NotImplementedError()
 
     @abc.abstractmethod
-    def on_sync(self, data):
+    async def on_sync(self, data):
```
Comment (richvdh) [inline synapse/replication/tcp/protocol.py]: the implementations of this don't seem to be async (and it's not `awaited` where it is called).

**Comment by erikjohnston** [comment]
Honestly its like I can't even tell the difference between ALL CAPS and lower case...

**Pull request merged**